linkerd/stack/src/failfast.rs

```diff
@@ -2,153 +2,300 @@
-                    State::Open => {
+        loop {
+            match mem::replace(&mut self.state, State::Invalid) {
+                State::Open(mut inner) => match inner.poll_ready(cx) {
+                    // The inner service just transitioned to `Pending`, so initiate a new timeout.
+                    Poll::Pending => {
                         self.wait
                             .as_mut()
                             .reset(Instant::now() + self.max_unavailable);
-                        debug!("{} service has become unavailable", self.scope);
-                        State::Waiting
+                        debug!("Service has become unavailable");
+                        self.state = State::Waiting(inner)
                     }
-
-                    // A timeout has been set, so wait for it to complete.
-                    State::Waiting => {
+                    Poll::Ready(res) => {
+                        self.state = State::Open(inner);
+                        return Poll::Ready(res.map_err(Into::into));
+                    }
+                },
+                State::Waiting(mut inner) => match inner.poll_ready(cx) {
+                    // The inner service became ready, transition back to `Open`.
+                    Poll::Ready(res) => {
+                        trace!("Service has become ready");
+                        self.state = State::Open(inner);
+                        return Poll::Ready(res.map_err(Into::into));
+                    }
+                    // The inner service hasn't become ready yet --- has the
+                    // timeout elapsed?
+                    Poll::Pending => {
                         let poll = self.wait.as_mut().poll(cx);
+
                         if poll.is_pending() {
-                            trace!("Pending");
+                            // The timeout hasn't elapsed yet, keep waiting.
+                            trace!("Service is Pending");
+                            self.state = State::Waiting(inner);
                             return Poll::Pending;
                         }
-                        warn!(
-                            "{} entering failfast after {:?}",
-                            self.scope, self.max_unavailable
-                        );
-                        State::FailFast
-                    }
 
-                    // Admit requests and fail them immediately.
-                    State::FailFast => {
-                        debug!("{} in failfast", self.scope);
+                        warn!("Service entering failfast after {:?}", self.max_unavailable);
+                        self.shared.in_failfast.store(true, Ordering::Release);
+
+                        let shared = self.shared.clone();
+                        self.state = State::FailFast(task::spawn(async move {
+                            let res = inner.ready().await;
+                            shared.exit_failfast();
```

Comment:
```suggestion
                            // Notify the paired `Gate` instances to begin
                            // advertising readiness so that the failfast
                            // service can advance. 
                            shared.exit_failfast();
```